core/src/token_info/cached.rs

```diff
@@ -60,14 +74,31 @@ impl TokenInfoCache {
 impl TokenInfoFetching for TokenInfoCache {
     fn get_token_info<'a>(&'a self, id: TokenId) -> BoxFuture<'a, Result<TokenBaseInfo>> {
         async move {
-            if let Some(info) = self.cache.read().await.get(&id).cloned() {
-                return Ok(info);
+            match self.cache.read().await.get(&id) {
+                Some(CacheEntry::TokenBaseInfo(info)) => return Ok(info.clone()),
+                Some(CacheEntry::UnretryableError(reason)) => {
+                    return Err(anyhow!("cached error: {}", reason));
+                }
+                None => (),
             }
             let info = self.inner.get_token_info(id).await;
-            if let Ok(info) = info.as_ref() {
-                self.cache.write().await.insert(id, info.clone());
+            match info {
+                Ok(info) => {
+                    self.cache
+                        .write()
+                        .await
+                        .insert(id, CacheEntry::TokenBaseInfo(info.clone()));
+                    Ok(info)
+                }
+                Err(err) if is_revert(&err) => {
+                    self.cache
```

Comment:
nit: Can we `log::debug!("{:?}")` the error here so we print the stack trace and context. Mostly because we demote the error into a `String` and lose this information.